src/HealthChecks.UI/Core/HostedService/HealthCheckReportCollectorHostedService.cs

```diff
@@ -49,20 +55,24 @@ private async Task ExecuteAsync(CancellationToken cancellationToken)
 
             while (!cancellationToken.IsCancellationRequested)
             {
-                _logger.LogDebug("Executing HealthCheck collector HostedService.");
-
-                using (var scope = scopeFactory.CreateScope())
+                // Collect should not be triggered until IServerAddressFeature reports the listening endpoints

```

Comment:
If we do not want a race condition where the hosted service trigger collections using relative urls and the webserver has not reported the listening urls yet we should control that the ServerAddressesService already know this urls

Reply:
@Tratcher @jkotalik @rynowak this is the part of the PR I am most concerned about as I do not know if they is another way of waiting until the listening urls are reported by the IServerAddressesFeature when using IIS with InProcess hosting model. 

Without this knowledge, I can just delay the healthcheck collection execution until the ServerAddressesService knows this listerning addresses.

Could you give some advice?

Reply:
@davidfowl helped figure out a workaround.

```C#
    public class AddressHostedService : IHostedService
    {
        private readonly IServer _server;
        private readonly IHostApplicationLifetime _lifetime;

        public AddressHostedService(IServer server, IHostApplicationLifetime lifetime)
        {
            _server = server;
            _lifetime = lifetime;
            _lifetime.ApplicationStarted.Register(() =>
            {
                var addresses = _server.Features.Get<IServerAddressesFeature>().Addresses.ToArray();
            });
        }
    }
```

The `ApplicationStarted` cancellation token will trigger once all of the hosted services have initialized. This way you can be notified once everything is *fully* started.

----

The problem in your case is that there's an order dependency between your hosted service and the "web host" hosted service. By default if you register a hosted service from `ConfigureServices` it's going to initialize before the web host does.

Reply:
@rynowak and @davidfowl thank you so much for your help!. I'm going to try it this way 👍

Reply:
@rynowak @Davidfowl, I still have the same race condition using ApplicationStarted because as you explained,  ApplicationStarted triggers when all the hosted services have started. And there is my problem, because the collector hosted service should not start until the webhost hosted service has started (the IIS problem), so I decided triggering the collection in the Register callback because I can't move our hosted service to the webhost configure services.

The following code is working ok and there is not further need to check if the host has addresses reported because they are always available when the callback triggers

Do you think this is correct?

```csharp
  private Task ExecuteAsync(CancellationToken cancellationToken)
  {
      _lifetime.ApplicationStarted.Register(async () => await Collect(cancellationToken));
      return Task.CompletedTask;
   }
```

Reply:
That's what I expected you would do, sorry if my example was unclear.

Reply:
As always, thank you very much @rynowak. Much appreciated!